File: lld/week-02-03-patterns/adapter/go/adapter_test.go
package adapter

import "testing"

func TestAdapterTranslatesLegacyCall(t *testing.T) {
	legacy := NewLegacyXMLDataProvider(map[string]string{"u1": "42"})
	adapted := NewXMLToJSONAdapter(legacy)

	rec, err := adapted.FetchJSON("u1")
	if err != nil {
		t.Fatalf("unexpected err: %v", err)
	}
	if rec["id"] != "u1" || rec["value"] != "42" {
		t.Fatalf("expected {id:u1 value:42}, got %v", rec)
	}
}

func TestAdapterPropagatesNotFound(t *testing.T) {
	legacy := NewLegacyXMLDataProvider(map[string]string{})
	adapted := NewXMLToJSONAdapter(legacy)

	if _, err := adapted.FetchJSON("missing"); err == nil {
		t.Fatal("expected error for missing record")
	}
}

func TestClientCodeIsProviderAgnostic(t *testing.T) {
	legacy := NewXMLToJSONAdapter(NewLegacyXMLDataProvider(map[string]string{"a": "10", "b": "20"}))
	modern := NewModernDataProvider(map[string]string{"a": "10", "b": "20"})

	legacySum, err := FetchAndSum(legacy, []string{"a", "b"})
	if err != nil {
		t.Fatalf("unexpected err: %v", err)
	}
	modernSum, err := FetchAndSum(modern, []string{"a", "b"})
	if err != nil {
		t.Fatalf("unexpected err: %v", err)
	}
	if legacySum != 30 || modernSum != 30 {
		t.Fatalf("expected both sums to be 30, got legacy=%d modern=%d", legacySum, modernSum)
	}
}

func TestLegacyProviderRawXML(t *testing.T) {
	legacy := NewLegacyXMLDataProvider(map[string]string{"u1": "42"})
	raw, err := legacy.FetchXML("u1")
	if err != nil {
		t.Fatalf("unexpected err: %v", err)
	}
	if raw == "" {
		t.Fatal("expected non-empty raw XML")
	}
}

func TestDemo(t *testing.T) {
	legacy := NewLegacyXMLDataProvider(map[string]string{"u1": "100"})
	var provider DataProvider = NewXMLToJSONAdapter(legacy)
	rec, err := provider.FetchJSON("u1")
	if err != nil {
		t.Fatalf("unexpected err: %v", err)
	}
	t.Logf("adapted record: %v", rec)
}

File: lld/week-02-03-patterns/adapter/java/src/AdapterTest.java
import java.util.List;
import java.util.Map;

/**
 * Plain assert-based test runner (no JUnit dependency needed).
 * Run via `java -cp out Main` (invoked from Main) or `java -cp out AdapterTest` directly.
 */
public class AdapterTest {

    public static void main(String[] args) throws Exception {
        runAll();
    }

    public static void runAll() throws Exception {
        testAdapterTranslatesLegacyCall();
        testAdapterPropagatesNotFound();
        testClientCodeIsProviderAgnostic();
        System.out.println("All AdapterTest cases passed.");
    }

    private static void testAdapterTranslatesLegacyCall() throws Exception {
        LegacyXmlDataProvider legacy = new LegacyXmlDataProvider(Map.of("u1", "42"));
        DataProvider adapted = new XmlToJsonAdapter(legacy);
        Map<String, String> record = adapted.fetchJson("u1");
        assertEquals("u1", record.get("id"), "adapted id");
        assertEquals("42", record.get("value"), "adapted value");
    }

    private static void testAdapterPropagatesNotFound() {
        LegacyXmlDataProvider legacy = new LegacyXmlDataProvider(Map.of());
        DataProvider adapted = new XmlToJsonAdapter(legacy);
        try {
            adapted.fetchJson("missing");
            throw new AssertionError("expected RecordNotFoundException");
        } catch (RecordNotFoundException e) {
            // expected
        }
    }

    private static void testClientCodeIsProviderAgnostic() throws Exception {
        DataProvider legacy = new XmlToJsonAdapter(new LegacyXmlDataProvider(Map.of("a", "10", "b", "20")));
        DataProvider modern = new ModernDataProvider(Map.of("a", "10", "b", "20"));

        int legacySum = Client.fetchAndSum(legacy, List.of("a", "b"));
        int modernSum = Client.fetchAndSum(modern, List.of("a", "b"));
        assertEquals(30, legacySum, "legacy-backed sum");
        assertEquals(30, modernSum, "modern-backed sum");
    }

    private static void assertEquals(Object expected, Object actual, String label) {
        if (!expected.equals(actual)) {
            throw new AssertionError(label + ": expected " + expected + " but got " + actual);
        }
    }
}
